### PR TITLE
Add Search Modal component

### PR DESCRIPTION
**Pull request by rodmachen**
Signed-off-by: Rod Machen <rmachen@expediagroup.com>

## Hey, I just made a Pull Request!

Per our demo in the community meetup, this is Expedia's contribution of a Modal to the roster of Search Components. For this first iteration, we'll just be featuring basic search, results, error page, and pagination.

Screen Shot 2021-09-24 at 9 41 24 AM
Screen Shot 2021-09-24 at 9 41 43 AM

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
Ooh, nice!

Remember to DCO sign all your commits (the build details of that error contains some details), and to run `yarn build:api-reports` when you change contracts and commit those results too.

**Review by emmaindal — COMMENTED** [review]
Awesome!! It would be great to have some tests for the new introduced Search Modal components. Want to add some?

File: plugins/search/src/components/SearchModal/SearchModal.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  container: {
+    borderRadius: 30,
+    display: 'flex',
+    height: '2.4em',
+  },
+  input: {
+    flex: 1,
+  },
+  paperFullWidth: { minHeight: 'calc(100% - 128px)' },
+  dialogActionsContainer: { padding: '8px 24px' },
```
Comment (camilaibs) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:50]: Maybe you can use the theme.spacing available for making styles:
```jsx
makeStyles(theme => ({
  dialogActionsContainer: {
    minHeight: theme.spacing(1, 3),
  },
}));
``

**Comment by rodmachen** [comment]
@emmaindal Absolutely. Tests are next up. I'm thinking we should also update docs?

File: plugins/search/src/components/SearchModal/SearchModal.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            onClear={handleClear}
+          />
+        </Paper>
+      </DialogTitle>
+      <DialogContent>
+        <SearchResult>
+          {({ results }) => (
+            <List>
+              {results.map(({ document }) => (
+                <div
```
[the remark this answers is not in the document]
Comment (rodmachen) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:106]: This is something we were considering for a future iteration, but if there's a call for having it now, we could possibly add it to this work. If not, we'll definitely be looking into it soon.

[the remark this answers is not in the document]
Comment (iamEAP) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:106]: +1 would be great to maybe navigate to the search page with the query/details filled out on `enter` key in the bar?  And perhaps a `see all results` link somewhere toward the bottom of the modal?

File: plugins/search/src/components/SearchModal/SearchModal.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+      }}
+      onClose={toggleModal}
+      aria-labelledby="customized-dialog-title"
+      open={open}
+      fullWidth
+      maxWidth="lg"
+    >
+      <DialogTitle id="customized-dialog-title">
+        <Paper className={classes.container}>
+          <SearchBarBase
```
Comment (iamEAP) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:93]: Would be great to maybe navigate to the search page with the query/details filled out on `enter` key in the bar?  And perhaps a `see all results` link somewhere toward the bottom of the modal?

So there's still a clear path to the search page for more advanced search needs.

File: plugins/search/src/components/SearchModal/SearchModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamEAP) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:106]: +1

File: plugins/search/src/components/SearchModal/SearchModal.tsx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            <List>
+              {results.map(({ document }) => (
+                <div
+                  role="button"
+                  tabIndex={0}
+                  key={`${document.location}-btn`}
+                  onClick={handleResultClick}
+                  onKeyPress={handleKeyPress}
+                >
+                  <DefaultResultListItem
```
Comment (iamEAP) [inline plugins/search/src/components/SearchModal/SearchModal.tsx:113]: This has me wondering how we might be able to enable App Integrators to customize/compose this.  What if I as an Integrator want it to use the TechDocs result list item for TechDocs results?  Or the Catalog result list item for Catalog results?

Not something we need to address in this PR, but something to think about for the future.

File: plugins/search/src/index.ts [hunk 5]
```diff
@@ -28,20 +28,22 @@ export {
   SearchContextProvider,
   SearchFilter,
   SearchFilterNext,
+  SearchModal,
   SearchPage as Router,
   SearchResultPager,
   SearchType,
   SidebarSearch,
+  SidebarSearchModal,
```
Comment (iamEAP) [inline plugins/search/src/index.ts:36]: I might recommend, rather than exporting the component directly here, that you export this as a component extension in `plugin.ts`.  Something like:

```tsx
export const SidebarSearchModal = searchPlugin.provide(
  createComponentExtension({
    component: {
      lazy: () =>
        import('./components/SidebarSearchModal').then(m => m.SidebarSearchModal),
    },
  }),
);
```

^ Gets us a nice error boundary, lazy loading, and more.

File: plugins/search/src/components/SearchModal/SearchModal.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (camilaibs) [reply plugins/search/src/components/SearchModal/SearchModal.tsx:106]: It's okay to be in the future

**Review by iamEAP — COMMENTED** [review]
Just summarizing a conversation with @rodmachen over discord:

1. This is _awesome_!  We're pumped to bring this in.
2. Love the iterative approach; let's keep track of definite future iterations (e.g. the keyboard navigation that @camilaibs mentioned, filtering, customization of the result components, etc) and create follow-up issues when this is merged.

Before merge, in addition to minor suggestions already on this PR, it would be great to address:
- [ ] Some way to navigate to the all results page from the modal (since this would take over the existing and only way to get there).
- [ ] Update `create-app` so that it uses this modal instead of the existing search bar (the existing change in `packages/app` only affects developers working in this monorepo)
- [ ] Some basic tests for the modal
- [ ] Updated documentation ([perhaps here](https://discord.com/channels/@me/841835285855862794/892067743224979497))
- [ ] Per @freben's comment: don't forget to make sure your commits are signed off appropriately.
- [ ] Also don't forget changesets for the plugins affected by this PR (should just be two `patch`-level changes, one for `@backstage/plugin-search` and one for `@backstage/create-app` once you make that create app change--check out this changeset as [a good template to follow](https://github.com/backstage/backstage/blob/9325075eeabb65f73b8fb1ce8cf57abe6377d30c/.changeset/brave-shoes-push.md) for the `create-app` one).

**Comment by emmaindal** [comment]
@iamEAP can I add a suggestion to your list of things? To add this component to storybook as well under the "Search" section, for visibility - [here is a example component ](https://backstage.io/storybook/?path=/story/plugins-search-defaultresultlistitem--default) and [its story file](https://github.com/backstage/backstage/blob/master/plugins/search/src/components/DefaultResultListItem/DefaultResultListItem.stories.tsx). Either we do it together with this PR, or if we want to continue iterate on the component - we can create a issue as an action!

**Comment by rodmachen** [comment]
@emmaindal That's a great idea. I'm going to have our designer open an issue and work on this after this PR is closed.

**Comment by emmaindal** [comment]
> @emmaindal That's a great idea. I'm going to have our designer open an issue and work on this after this PR is closed.

sounds good to me!

**Comment by rodmachen** [comment]
PR updated. The `View All Results` button/link is still in progress.

**Comment by emmaindal** [comment]
> PR updated. The `View All Results` button/link is still in progress.

Nice, the changes looks good to me, looking at the checklist above there are just a few things left. 

- [ ] Some way to navigate to the all results page from the modal (since this would take over the existing and only way to get there). 

^ you mentioned that already. 

- [ ] Updated documentation @iamEAP your link takes me to discord, did you think of any specific place to add docs?
(otherwise I'm suggesting somewhere here https://backstage.io/docs/features/search/getting-started#adding-search-to-the-frontend)

- [ ]  Per @freben's comment: don't forget to make sure your commits are signed off appropriately.
(if you go to the failed DCO step you get instructions for how to do it on the commits missing)


Storybook we can do in a separate PR as you suggested!

**Comment by emmaindal** [comment]
> navigate to the all results page, docs, and sign-off have been updated. Please review again.
> 
> Tests are passing locally, but not on the GitHub check. Unsure as to why.

Awesome @rodmachen will look at your changes. The reason why the build doesn't pass is because api reports I think need to get re-generated for these changes. Can you try to run `yarn build:api-reports` and commit the updates!

**Comment by rodmachen** [comment]
@emmaindal that fixed some checks, but some still broken. Please advise if you have a fix.

**Comment by rodmachen** [comment]
Updated. Please take a look @iamEAP @emmaindal .

**Comment by emmaindal** [comment]
> Updated. Please take a look @iamEAP @emmaindal .

Hey @rodmachen we are so close to get this 🚢, I see some CI and E2E tests gives typescript errors which seems to be unrelated to your changes. Can you try rebase this on top of master and see if that solves it?

File: packages/catalog-client/package.json [hunk 6]
```diff
@@ -2,8 +2,8 @@
   "name": "@backstage/catalog-client",
   "description": "An isomorphic client for the catalog backend",
   "version": "0.5.0",
-  "main": "src/index.ts",
-  "types": "src/index.ts",
+  "main": "dist/index.cjs.js",
+  "types": "dist/index.d.ts",
```
Comment (emmaindal) [inline packages/catalog-client/package.json:6]: what are those changes required for?

File: packages/catalog-client/package.json [hunk 7]
```diff
@@ -41,5 +41,6 @@
   },
   "files": [
     "dist"
-  ]
+  ],
+  "module": "dist/index.esm.js"
```
Comment (emmaindal) [inline packages/catalog-client/package.json:45]: same as above, what is this change required for?

File: packages/catalog-client/package.json (hunk 6 above)
Comment (rodmachen) [reply packages/catalog-client/package.json]: I didn't make these changes intentionally, so maybe it's part of some linting/cleaning process.

Regardless, I'm changing it to what's in master. I'm a little worried that this didn't happen during rebase.

File: packages/catalog-client/package.json (hunk 7 above)
Comment (rodmachen) [reply packages/catalog-client/package.json]: same as above

**Review by emmaindal — APPROVED** [review]
Awesome!! All green and a ✅  from me! Thanks a lot for this contribution @rodmachen 🎉

**Review by freben — APPROVED** [review]
Looks good to me too!

**Comment by freben** [comment]
Haha, sorry for some late feedback because I played around with it a bit more - some notes for the future:

- When click search on the sidebar, my intent is definitely to start typing, so the input box should probably have focus when the modal opens.
- One whole second of typing debounce is perhaps a bit much - it started to feel sluggish after a while, probably lower that to something like a quarter of a second or so.
- When I press View Full Results, it should also close the modal.
- On a (vertically) small screen, the View Full Results link scrolls away as I scroll down among the results, it should probably stay put (or be a button inside the search box perhaps, next to the clear button?).
- The Next and Back buttons should probably be Previous and Next.
- The Next and Back buttons have almost zero padding (making them very small click targets) but their container has a huge padding, making their edges not align with other content as expected. Should probably use a plain unstyled `<DialogActions>` there, or at most giving it a bit of `paddingTop` internally if they get too close to the results list.

Anyway, great contribution, let's iterate!

**Comment by emmaindal** [comment]
> Haha, sorry for some late feedback because I played around with it a bit more - some notes for the future:
> 
> * When click search on the sidebar, my intent is definitely to start typing, so the input box should probably have focus when the modal opens.
> * One whole second of typing debounce is perhaps a bit much - it started to feel sluggish after a while, probably lower that to something like a quarter of a second or so.
> * When I press View Full Results, it should also close the modal.
> * On a (vertically) small screen, the View Full Results link scrolls away as I scroll down among the results, it should probably stay put (or be a button inside the search box perhaps, next to the clear button?).
> * The Next and Back buttons should probably be Previous and Next.
> * The Next and Back buttons have almost zero padding (making them very small click targets) but their container has a huge padding, making their edges not align with other content as expected. Should probably use a plain unstyled `<DialogActions>` there, or at most giving it a bit of `paddingTop` internally if they get too close to the results list.
> 
> Anyway, great contribution, let's iterate!

great feedback @freben, we are planning to open a PR adding this component to storybook so I can go on and do that + fix your suggestions! Thanks!
